ipwb/reconstructive.js

```diff
@@ -0,0 +1,371 @@
+   * @return  {boolean}           - Should the request be rerouted?
+   */
+  function shouldExclude(event, config) {
+    return Object.keys(exclusions).some(key => {
+      if (exclusions[key](event, config)) {
+        config.debug && console.log('Exclusion found:', key, event.request.url);
+        return true;
+      }
+      return false;
+    });
+  }
+
+  /**
+   * createUrim - Creates a potential URI-M based on the requested URL and the referrer URL for request rerouting.
+   *
+   * @private
+   * @param   {FetchEvent} event  - The fetch event.
+   * @return  {string}            - A potential URI-M.
+   */
+  function createUrim(event) {
+    // Extract datetime and the URI-R of the referrer.
+    let [datetime, refUrir] = extractDatetimeUrir(event.request.referrer);
+    let urir = new URL(event.request.url);
+    // This condition will match when the request was initiated from an absolute path and fail if it was an absolute URL.
+    if (urir.origin == self.location.origin) {
+      // If it was an absolute path then referrer's origin was used.
+      // We need to replace it with the origin of the referrer's URI-R instead.
+      // The RegExp used will capture the origin with the protocol, if any (http, https, or BLANK).
+      let refOrigin = refUrir.match(/^(https?:\/\/)?[^\/]+/)[0];
+      urir = refOrigin + urir.pathname + urir.search;
+    } else {
+      urir = urir.href;
+    }
+    return config.urimPattern.replace('<datetime>', datetime).replace('<urir>', urir);
+  }
+
+  /**
+   * extractDatetimeUrir - Extracts datetime and URI-R from a URI-M.
+   *
+   * @private
+   * @param   {string} urim - A URI-M.
+   * @return  {array}       - An array of datetime and URI-let R.
+   */
+  function extractDatetimeUrir(urim) {
+    let [, datetime, urir] = urim.match(config.urimRegex);
+    // Swap the two extracted values if the datetime pattern appeared after the URI-R.
+    // This is not a common practice, but possible if an archive uses query parameters instead of paths.
+    if (isNaN(datetime)) {
+      [datetime, urir] = [urir, datetime];
+    }
+    return [datetime, urir];
+  }
+
+  /**
+   * createRequest - Creates a new request based on the original.
+   *                 Coppies all the headers from the original request.
+   *                 Adds X-ServiceWorker header with the id of the module.
+   *                 Sets the redirect mode to manual to ensure proper origin boundaries.
+   *
+   * @private
+   * @param   {FetchEvent} event  - The fetch event.
+   * @return  {Request}           - A new request object.
+   */
+  function createRequest(event) {
+    let headers = cloneHeaders(event.request.headers);
+    headers.set('X-ServiceWorker', config.id);
+    return new Request(event.request.url, {headers: headers, redirect: 'manual'});
+  }
+
+  /**
+   * cloneHeaders - Clones provided request or response headers.
+   *
+   * @private
+   * @param   {Headers} original - Origina request or response headers.
```

Comment:
"Origina" missing L in comment, not a big issue.

Reply:
Will fix it in the upstream Reconstructive repo and bring that change with the upcoming banner related PR.